### PR TITLE
Export Data working

### PR DESCRIPTION
**Pull request by anthony2261**
- [X] Stream results BE -> FE
- [X] Stream results SQL -> BE
- [ ] Validate that we can export the data for the given result_id

### CODE REVIEW

File: backend/dataline/services/llm_flow/utils.py [hunk 1]
```diff
@@ -109,6 +109,29 @@ def from_uri(
         engine = create_engine(database_uri, **_engine_args)
         return cls(engine, schemas=schemas, **kwargs)
 
+    def custom_run_sql_stream(self, query: str) -> Generator[Sequence[Row[Any]], Any, None]:
+        # https://docs.sqlalchemy.org/en/20/core/connections.html#streaming-with-a-fixed-buffer-via-yield-per
+        yield_per = 1000
+        if self.dialect == "mssql":
```
Comment (RamiAwar) [inline backend/dataline/services/llm_flow/utils.py:115]: This is becoming a little annoying seeing all these if-statements everywhere. Idk if it's time for this yet, cause it's still acceptable, but we should at least start thinking about using some better design pattern here, ex. Command pattern or something where based on the Dialect, the functionality is housed in one place.

Comment (RamiAwar) [inline backend/dataline/services/llm_flow/utils.py:115]: This is becoming a little annoying seeing all these if-statements everywhere. Idk if it's time for this yet, cause it's still acceptable, but we should at least start thinking about using some better design pattern here, ex. Command pattern or something where based on the Dialect, the functionality is housed in one place.

File: backend/dataline/api/result/router.py [hunk 2]
```diff
@@ -39,3 +40,25 @@ async def refresh_chart_result_data(
...
+
+
+@router.get("/result/{result_id}/export")
+async def export_results(
+    result_id: UUID,
+    session: Annotated[AsyncSession, Depends(get_session)],
+    result_service: Annotated[ResultService, Depends(ResultService)],
+    background_tasks: BackgroundTasks,
+) -> StreamingResponse:
+    background_tasks.add_task(posthog_capture, "results_exported")
```
[the remark this answers is not in the document]
Comment (RamiAwar) [reply backend/dataline/api/result/router.py:52]: Same with function/route name maybe

File: backend/dataline/api/result/router.py [hunk 3]
```diff
@@ -39,3 +40,25 @@ async def refresh_chart_result_data(
...
+    try:
+        return await result_service.export_results_as_csv(session, result_id)
+    except ValueError as e:
+        raise HTTPException(status_code=400, detail=str(e))
+
+
+@router.get("/result/{result_id}/export-url")
+async def export_csv_url(request: Request, result_id: UUID) -> SuccessResponse[str]:
```
Comment (RamiAwar) [inline backend/dataline/api/result/router.py:60]: Don't understand why we need this

File: backend/dataline/services/result.py [hunk 4]
```diff
@@ -118,3 +115,37 @@ async def validate_sql_query_result_for_chart(
...
+        # Parse the SQL query from the content
+        query_content = SQLQueryStringResultContent.model_validate_json(query_string_result.content)
+        sql_query = query_content.sql
+
+        # Get the connection for the result
+        connection = await self.result_repo.get_connection_from_result(session, result_id)
+        db = SQLDatabase.from_dataline_connection(Connection.model_validate(connection))
+
+        # Create a generator function to stream the results
+        async def generate_csv():
```
Comment (RamiAwar) [inline backend/dataline/services/result.py:134]: I would extract this into a separate function that takes as input a generator and creates a csv generator from it.

Not a fan of in-line function definitions, unless it really makes sense. I would still separate it even if this is only used in a single place. Readability counts more than most things imo

File: frontend/src/components/Library/DynamicTable.tsx [hunk 5]
```diff
@@ -103,6 +107,17 @@ export const DynamicTable: React.FC<{
                 </button>
               </CustomTooltip>
             ))}
+          {data.rows.length > 1 && linked_id && (
+            <CustomTooltip hoverText="Export">
```
Comment (RamiAwar) [inline frontend/src/components/Library/DynamicTable.tsx:111]: "Export as CSV"?

File: backend/dataline/services/llm_flow/utils.py (hunk 1 above)
Comment (anthony2261) [reply backend/dataline/services/llm_flow/utils.py:115]: Resolving - Duplicate comment

File: backend/dataline/api/result/router.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (anthony2261) [reply backend/dataline/api/result/router.py]: Done

File: backend/dataline/api/result/router.py (hunk 3 above)
Comment (anthony2261) [reply backend/dataline/api/result/router.py]: Removed

File: frontend/src/components/Library/DynamicTable.tsx (hunk 5 above)
Comment (anthony2261) [reply frontend/src/components/Library/DynamicTable.tsx:111]: Keeping it as Export

**Comment by anthony2261** [comment]
@RamiAwar made required changes as well as ones discussed in yesterday's call (using anchor tag to open the download in a new tab...). If you have no comments after reviewing, merge away

**Review by RamiAwar — APPROVED** [review]

**Pull request merged**